### PR TITLE
adding katacoda tutorials

### PR DESCRIPTION
**Pull request by sbawaska**
katacoda scenarios let people try out Knative without installing anything on their local machines. The goal is add a number of basic as well as advanced scenarios that would serve as a guided tutorial for knative. This PR adds four total scenarios:
1. Getting started with Serving using kn cli
2. Getting started with Serving using yaml 
3. Getting started with Eventing with Channels
4. Getting started with Eventing with Broker

You can run these tutorials at: https://www.katacoda.com/knative-tutorials/

### CODE REVIEW

**Comment by sbawaska** [comment]
Screen Shot 2021-01-12 at 9 50 22 PM Once merged, I will create a PR to the website repo as well.

**Comment by abrennan89** [comment]
@sbawaska I love this, thank you for taking the time to add it.
I've added @RichieEscarez to the review from a website perspective, I'll try to take a look at the content this week.

File: tutorials/katacoda/1-serving-intro/index.json
Comment (mpetason) [inline tutorials/katacoda/1-serving-intro/index.json:3]:
```suggestion
  "description": "Introduction to Knative by installing knative serving and deploying an application (If you want to script your workflow, or really don't like yaml)",
```

File: tutorials/katacoda/1-serving-intro/intro.md
Comment (mpetason) [inline tutorials/katacoda/1-serving-intro/intro.md:2]:
```suggestion
Knative brings the "serverless" experience to kubernetes. It also tries to codify common patterns and best practices for running applications while hiding the complexity of doing so on kubernetes. It does so by providing two components:
```

File: tutorials/katacoda/2-serving-intro-yaml/intro.md
Comment (mpetason) [inline tutorials/katacoda/2-serving-intro-yaml/intro.md:2]:
```suggestion
Knative brings the "serverless" experience to kubernetes. It also tries to codify common patterns and best practices for running applications while hiding the complexity of doing so on kubernetes. It does so by providing two components:
```

File: tutorials/katacoda/3-eventing-intro-channels/step3.md
Comment (mpetason) [inline tutorials/katacoda/3-eventing-intro-channels/step3.md:9]:
```suggestion
problems (like what if the producer crashes after delivering an event to only a subset of consumers?
```

**Review by mpetason — COMMENTED** [review]
Initial review - I need to take some time and read through the rest again.

File: tutorials/katacoda/1-serving-intro/step1.md
Comment (csantanapr) [inline tutorials/katacoda/1-serving-intro/step1.md:6]: harcoding the version would be hard to maintain `v0.19.0` is located in 23 places in this PR

@abrennan89 Do you know if there is a variable we can use per branch to extract the `0.19` ?

Comment (abrennan89) [reply tutorials/katacoda/1-serving-intro/step1.md:6]: @markusthoemmes is probably a better person to ask

Comment (sbawaska) [reply tutorials/katacoda/1-serving-intro/step1.md:6]: elsewhere in the docs, the version number is referred like this: `kubectl apply --filename {{< artifact repo="serving" file="serving-crds.yaml" >}}`. That works great in the tooling used by the website, however,  these files are going to be consumed by katacoda, hence it is not possible to use that same syntax. 

How do we feel about updating these versions as part of the release process?

Comment (csantanapr) [reply tutorials/katacoda/1-serving-intro/step1.md:6]: Yeah that a possibility to update for each release cut, but would need to be automated with the scripts that cut a release or update api ref spec @RichieEscarez migjt have guidance on location of scripts

**Comment by abrennan89** [comment]
Folks this is not something I will have bandwidth to review any time soon, and I'm not sure that Katacoda tutorials content reviews should really be part of Docs WG scope anyway, since they're kind of their own thing outside of our usual written markdown content.

@omerbensaadon is this something you can look at as part of the UX WG instead maybe?

File: tutorials/katacoda/1-serving-intro/step1.md
Comment (csantanapr) [reply tutorials/katacoda/1-serving-intro/step1.md:6]: I still see issues having this pointing to an old version of knative. For example we cut version 0.21

Perhaps what do you think if we don't hardcore a specific version and instead we use the nightly URL instead?

Comment (csantanapr) [reply tutorials/katacoda/1-serving-intro/step1.md:6]: Oh wait this is running on Linux not users computer, I proposed you use bash/sh for example cusinf curl to find out the latest cut version and stored that in variable then use the variable when installing with kubectl URL

**Review by csantanapr — CHANGES_REQUESTED** [review]
This PR is a good idea, but right now is block until it get's updated to remove the hardcoded version of v0.19. It needs to be a bash script that fetches what's the latest versioned released from github released, and uses that version.

**Comment by abrennan89** [comment]
/hold

Blocked as per Carlos's comment

File: tutorials/katacoda/1-serving-intro/step1.md
Comment (csantanapr) [inline tutorials/katacoda/1-serving-intro/step1.md:11]: I would prefer to use kourier instead of contour, less required resources

**Comment by sbawaska** [comment]
> This PR is a good idea, but right now is block until it get's updated to remove the hardcoded version of v0.19. It needs to be a bash script that fetches what's the latest versioned released from github released, and uses that version.

Hi @csantanapr , the hardcoded versions have been removed. Please review.

**Comment by omerbensaadon** [comment]
missed the assignment here but it looks like @csantanapr had me covered 🥇

**Review by omerbensaadon — COMMENTED** [review]
Would be a great addition to the site! 

Have we narrowed in on where this would live on the site?

**Comment by sbawaska** [comment]
> Would be a great addition to the site!
> 
> Have we narrowed in on where this would live on the site?

I was thinking this: https://github.com/knative/website/pull/246

File: tutorials/katacoda/3-eventing-intro-channels/step4.md
Comment (mpetason) [inline tutorials/katacoda/3-eventing-intro-channels/step4.md:49]:
```suggestion
Now, let's create the Sequence to execute the service "first" and then pass its result to the service "event-display-chain"
```

File: tutorials/katacoda/4-eventing-intro-broker/step2.md
Comment (mpetason) [inline tutorials/katacoda/4-eventing-intro-broker/step2.md:36]:
```suggestion
Now we will create consumers that will simply log the event. (We will only create two consumers, the rest will be very similar):
```

File: tutorials/katacoda/4-eventing-intro-broker/step2.md
Comment (mpetason) [inline tutorials/katacoda/4-eventing-intro-broker/step2.md:4]:
```suggestion
consuming/transforming which are then processed by even more functions and so on. It can get a bit unwieldy to keep track of which Channel is having which events. Also, sometimes you might only want to
consume specific types of events. You would have to receive all the events, and throw out the ones you’re not interested in. To make these kinds of interactions easier and allow the
```

**Review by mpetason — COMMENTED** [review]
This looks great! I added a few ideas that may read a little bit better plus a few small updates. Great work.

**Comment by sbawaska** [comment]
> This looks great! I added a few ideas that may read a little bit better plus a few small updates. Great work.

Thanks for the review @mpetason. I have incorporated your feedback.

File: tutorials/katacoda/1-serving-intro/step2.md
Comment (csantanapr) [inline tutorials/katacoda/1-serving-intro/step2.md:10]:
```suggestion
kn service create demo --image gcr.io/knative-samples/helloworld-go --env TARGET=blue --autoscale-window 15s
```

what do you think of specify a 15 seconds, this way the user doesn't have to seat there waiting for 60 seconds

File: tutorials/katacoda/1-serving-intro/step2.md
Comment (csantanapr) [inline tutorials/katacoda/1-serving-intro/step2.md:7]: Do you want to mention something that this step will take up to a minute, at least when I try it is taking about 1 minute?
Is to avoid the user thinking that something is wrong.

Comment (csantanapr) [reply tutorials/katacoda/1-serving-intro/step2.md:7]:
```
$ kn service create demo --image gcr.io/knative-samples/helloworld-go --env TARGET=blue
Creating service 'demo' in namespace 'default':

  0.155s The Route is still working to reflect the latest desired specification.
  0.218s Configuration "demo" is waiting for a Revision to become ready.
 50.677s ...
 50.816s Ingress has not yet been reconciled.
 50.874s Waiting for Envoys to receive Endpoints data.
 51.110s Waiting for load balancer to be ready
 51.329s Ready to serve.
 ```

File: tutorials/katacoda/1-serving-intro/step2.md
Comment (csantanapr) [inline tutorials/katacoda/1-serving-intro/step2.md:20]: It would be great to configure the minikube to use a local domain name such as `http://demo.default.127.0.0.1.nip.io`
Or a local IP `10.` when I use minikube I use `minikube tunnel` to get a local tunnel on an IP address for a load balancer IP.

File: tutorials/katacoda/1-serving-intro/step2.md
Comment (csantanapr) [inline tutorials/katacoda/1-serving-intro/step2.md:24]: When I click `watch kubectl get pods` It opens the terminal 2 in the right, side but nothing shows up.

Need to tell the user what to do to make the command output to show up
image

Comment (csantanapr) [reply tutorials/katacoda/1-serving-intro/step2.md:24]: I ad to click twice, maybe one is to focus the T2 terminal and the second one is to run the command, not sure

File: tutorials/katacoda/1-serving-intro/step3.md
Comment (csantanapr) [inline tutorials/katacoda/1-serving-intro/step3.md:2]: This section might be confusing for new users.

- Title mentions "Blue/Green" but there is not rollout from current version (blue) to the new version (green)
- The traffic is left at 50/50 traffic split, the new version never goes to 100%
- A tag is used in the example but the tag is never use, the tag is for the purpose of unique URL, but it was never use
- There is a command that feels like brack magic `kn revision list -o 'jsonpath={.items[0].metadata.name}`

Here is a [tutorial](https://github.com/csantanapr/knative-tekton#33-knative-service-traffic-splitting) with a flow that is more appropriate. 

We could use a revision name called it revision `blue`, and a revision `green` green being latest.
You can show that green version gets 50% traffic and blue 50%, and then show blue takes 0% and green 100%

File: tutorials/katacoda/2-serving-intro-yaml/step3.md
Comment (csantanapr) [inline tutorials/katacoda/2-serving-intro-yaml/step3.md:2]: I have similar feedback I gave back in [the cli section](https://github.com/knative/docs/pull/3142/files#r599149635)

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (csantanapr) [inline tutorials/katacoda/3-eventing-intro-channels/step2.md:7]: Really like the diagram pictures ❤️ 

Wondering if we can make diagrams for serving tutorial

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (csantanapr) [inline tutorials/katacoda/3-eventing-intro-channels/step2.md:88]: I saw problems with this execute, I had to click twice to make finally show up

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (csantanapr) [inline tutorials/katacoda/3-eventing-intro-channels/step2.md:85]: I like the approach of the wait for condition and then show the logs

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (csantanapr) [inline tutorials/katacoda/3-eventing-intro-channels/step2.md:87]: The logs don't show up on first run I click.
Maybe we need to put some type of sleep.

I had to run it twice to see the logs

File: tutorials/katacoda/3-eventing-intro-channels/step4.md
Comment (csantanapr) [inline tutorials/katacoda/3-eventing-intro-channels/step4.md:53]: Update all the `v1beta1` to `v1`

File: tutorials/katacoda/3-eventing-intro-channels/step4.md
Comment (csantanapr) [inline tutorials/katacoda/3-eventing-intro-channels/step4.md:101]: I had to run it twice to see the output of the logs

there was a timeout error also

```
pingsource.sources.knative.dev/ping-source-sequence created
$ # it is likely that is pod is still being created after scaling down to zero
$ kubectl wait --for=condition=ready pod -l serving.knative.dev/service=event-display-chain
pod/event-display-chain-00001-deployment-566755999b-g4xqg condition met
$ # see the logs
$ kubectl -n default logs -l serving.knative.dev/service=event-display-chain -c user-container --since=10m --tail=50
$ # it is likely that is pod is still being created after scaling down to zero
$ kubectl wait --for=condition=ready pod -l serving.knative.dev/service=event-display-chain
error: timed out waiting for the condition on pods/event-display-chain-00001-deployment-566755999b-g4xqg
$ # see the logs
$ kubectl -n default logs -l serving.knative.dev/service=event-display-chain -c user-container --since=10m --tail=50
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: dev.knative.sources.ping
  source: /apis/v1/namespaces/default/pingsources/ping-source-sequence
  id: d65624c9-70dd-4bbd-8ec0-d1bfbfbb39dd
  time: 2021-03-23T01:59:00.103779523Z
  datacontenttype: application/json
Data,
  {
    "id": 0,
    "message": "Hello Sequence! - Updating this message a little..."
  }
```

File: tutorials/katacoda/4-eventing-intro-broker/step2.md
Comment (csantanapr) [inline tutorials/katacoda/4-eventing-intro-broker/step2.md:56]: Is kind odd to start with consumer2 and then consumer1, I would expect you do them in order starting with something named consumer1

File: tutorials/katacoda/4-eventing-intro-broker/step3.md
Comment (csantanapr) [inline tutorials/katacoda/4-eventing-intro-broker/step3.md:29]: add break line before last sentence.

Also code quotes for the colors `red` `blue` `orange`

File: tutorials/katacoda/4-eventing-intro-broker/step3.md
Comment (csantanapr) [inline tutorials/katacoda/4-eventing-intro-broker/step3.md:77]: This is not working for `consumer2`

```
$ # see the logs
$ kubectl logs -l serving.knative.dev/service=consumer1 -c user-container --since=10m --tail=50
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: orange
  source: kata-tutorial
  id: test
  datacontenttype: application/json
Extensions,
  knativearrivaltime: 2021-03-23T02:14:43.808385659Z
Data,
  {
    "msg": "The event is orange!"
  }
$ kubectl logs -l serving.knative.dev/service=consumer2 -c user-container --since=10m --tail=50
$ # it is likely that is pod is still being created after scaling down to zero
$ kubectl wait --for=condition=ready pod -l serving.knative.dev/service=consumer1
pod/consumer1-00001-deployment-5b5b8c849d-cb8gh condition met
$ kubectl wait --for=condition=ready pod -l serving.knative.dev/service=consumer2
error: no matching resources found
$ # see the logs
$ kubectl logs -l serving.knative.dev/service=consumer1 -c user-container --since=10m --tail=50
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: orange
  source: kata-tutorial
  id: test
  datacontenttype: application/json
Extensions,
  knativearrivaltime: 2021-03-23T02:14:43.808385659Z
Data,
  {
    "msg": "The event is orange!"
  }
$ kubectl logs -l serving.knative.dev/service=consumer2 -c user-container --since=10m --tail=50
```

**Comment by csantanapr** [comment]
@sbawaska I did a review. 
Please take a at the comments and also the errors I saw while trying the tutorial

File: tutorials/katacoda/1-serving-intro/step3.md
Comment (sbawaska) [reply tutorials/katacoda/1-serving-intro/step3.md:2]: This was super helpful. I have incorporated this. Thanks!

File: tutorials/katacoda/1-serving-intro/step2.md
Comment (sbawaska) [reply tutorials/katacoda/1-serving-intro/step2.md]: Yes, I have seen it sometimes too. It probably is a bug in katacoda, I added a note to click twice.

File: tutorials/katacoda/1-serving-intro/step2.md
Comment (sbawaska) [reply tutorials/katacoda/1-serving-intro/step2.md]: added a note.

File: tutorials/katacoda/1-serving-intro/step2.md
Comment (sbawaska) [reply tutorials/katacoda/1-serving-intro/step2.md:20]: I think we can fix that in a follow on PR.

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (sbawaska) [reply tutorials/katacoda/3-eventing-intro-channels/step2.md:7]: Thanks! Will consider in a follow on PR

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (sbawaska) [reply tutorials/katacoda/3-eventing-intro-channels/step2.md:55]: Thanks!

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (sbawaska) [reply tutorials/katacoda/3-eventing-intro-channels/step2.md:85]: Thanks!

File: tutorials/katacoda/3-eventing-intro-channels/step4.md
Comment (sbawaska) [reply tutorials/katacoda/3-eventing-intro-channels/step4.md]: done!

File: tutorials/katacoda/4-eventing-intro-broker/step3.md
Comment (sbawaska) [reply tutorials/katacoda/4-eventing-intro-broker/step3.md]: done!

File: tutorials/katacoda/4-eventing-intro-broker/step2.md
Comment (sbawaska) [reply tutorials/katacoda/4-eventing-intro-broker/step2.md:56]: Good point. I have consumer1 creating two triggers, hence wanted to start with consumer2 first. Is it ok to fix this later?

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (sbawaska) [reply tutorials/katacoda/3-eventing-intro-channels/step2.md:88]: problem with execute might be a browser thing? I don't see issues with chrome. What browser are you using?

File: tutorials/katacoda/3-eventing-intro-channels/step4.md
Comment (sbawaska) [reply tutorials/katacoda/3-eventing-intro-channels/step4.md:101]: The default timeout was 30 seconds, bumped it to 90.

File: tutorials/katacoda/4-eventing-intro-broker/step3.md
Comment (sbawaska) [reply tutorials/katacoda/4-eventing-intro-broker/step3.md:79]: The default timeout was 30 seconds, bumped it to 90.

File: tutorials/katacoda/3-eventing-intro-channels/step2.md
Comment (sbawaska) [reply tutorials/katacoda/3-eventing-intro-channels/step2.md:87]: The default timeout was 30 seconds, bumped it to 90.

File: tutorials/katacoda/2-serving-intro-yaml/step3.md
Comment (sbawaska) [reply tutorials/katacoda/2-serving-intro-yaml/step3.md:2]: fixed!

**Comment by sbawaska** [comment]
> @sbawaska I did a review.
> Please take a at the comments and also the errors I saw while trying the tutorial

Thanks for the thorough review @csantanapr! I have incorporated most of it. please review. Thanks!

**Comment by csantanapr** [comment]
Hi @sbawaska 

You mentioned that you preferred to fix some items in a subsequent PR, would you mind collecting a these in an issue so you or others contributors could improve once this is merge?

I would like to contribute to it as part of new initiatives in the UX WG
cc @omerbensaadon

**Comment by omerbensaadon** [comment]
@sbawaska we are planning to test this in User Interviews beginning early next week! 

Not a deadline, we can test the basic premise (Getting Started w/ katacoda) with what we have, just thought I would let you know.

**Pull request merged**
